### ISSUE TITLE
Support switching between SX and MX variables

### ISSUE THREAD

**Issue by 4flixt** [do-mpc/do-mpc#34]
Especially for neural network system models, we need MX variables. 
Switching should be implemented in e.g. the model creation.

**Comment by Filippo-Tamagnini** [comment]
What are the implications of using MX symbolics in the model definition? how much of the do-mpc code would be affected?

**Comment by 4flixt** [comment]
Hi Filipo, thank you for your question. The change is surprisingly major. The model will need an overhaul. For example, it would be required to introduce all variables of a given type, e.g. states at once. CasADi does not support adding MX variables to an existing symbolic structure. The current implementation is based on that.

So we have to decide how the new API should look like, if it should replace the new API or be an alternative.

Since MX is really only beneficial for large matrix vector multiplications (in terms of execution speed), I have postponed this change to the next version.

**Comment by Filippo-Tamagnini** [comment]
Until then, do you think there is some feasible workaround to work with? 
I have a model that uses both SX and MX (interfaced through casadi functions and callbacks). Would it be possible to override the do-mpc variable definition and directly provide the MX expressions?

**Comment by 4flixt** [comment]
I think you would need to change all SX in the code to MX. Certainly that is not possible from the user API atm. Of course, you can clone and modify the source.
I assume it will be rather messy, however.

You are certain you need MX for your model? We found it only accelerates the code for large matrix-vector multiplications, e.g. using a feedforward neural network as the system model.

I am also wondering when and how you are mixing SX and MX in your model?

In any case, I am happy to hear that there is demand for this feature and can try to prioritize it in the future.

**Comment by Filippo-Tamagnini** [comment]
My use case is particular, as I'm using an algorithm that is not expressible with elementary operations. CasADi allows this kind of user defined function by subclassing the Callback class, however the generated function only supports MX symbolic arguments (or DM).

As for mixing SX and MX, this can be done through CasADi functions and might be used for example to define the state dependent function in a PDE model.
For example, i could write the reaction rate for a reaction in terms of the local state:
```
x_c = SX.sym("c");
kr = 0.5;
r = SX.sym("r");
r_F = Function("r_F",[x_c],[kr*x_c]);
```
And then to insert this in a distributed parameter model, simply by calling the scalar function defined above for a vector of states

```
x = MX("x",(N_grid,N_states));
dx_dt = MX("dx_dt",(N_grid,N_states));
Discr_Mat = DM((N_grid,N_grid)); #to be defined;
dx_dt[:,state_c] = Discr_Mat@x[state_c,:] + r_F(x[:,state_c])
```
This simple example could also be written in terms of SX variables, of course, but it just serves as a demonstration of how one can mix the two symbolic types.

I'll try the messy road, maybe i will also come up with a more general statement of the problem!

**Comment by 4flixt** [comment]
Ah you are working with a PDE system. I see how the MX variables would be better in that case. I also understand now what you mean by mixing SX and MX. I was confused, because even simple things like:

```python
x = SX.sym('x')
y = MX.sym('y')

x+y
```

will throw an error. 
So in terms of the choice MX/SX in your code above, the relevant section for CasADi and do-mpc is in MX variables.
If you come up with a better model configuration while working on this please feel free to share it! I'll also look into it soon, because I'll expect to work with MX in do-mpc in the near future (hence this issue).

**Comment by alexandru-tc** [comment]
When you first asked me about this, Filippo, I didn't quite understand your goal either. To be entirely honest, I still don't... So if you would like to give us some more details about your particular case, I would appreciate that. But I don't think this thread is the right place, we can have a separate conversation :) 

The example you are giving above is what I had in mind first, and also what I had also implemented previously for a multiple-shooting NMPC tutorial (don't know if I ever gave you the example script). CasADi even recommends to gather all constraints in an MX vector. Since the NLP takes the constraints as a vector, it is natural to implement the model as a vector of SX expressions and then use integrator calls to create an SX/MX vector of constraints.  Currently do-mpc defines it as SX, but that's not the problem. As long as each entry in the right hand side is a symbolic expression, we're good. 

If I understand it correctly, the problem comes from wanting to define the system of ODEs differently, all at once and as a matrix. Is that so? If yes,  I'm curious to understand what brings this about and what are the advantages. Maybe it has to do with coding aspects: vector matrix multiplications VS an iterative coding structure to generate all the ODE entries. As it stands, regardless of how the user prefers to implement the model, in order to use the do-mpc API it is required to have access to each entry in the model as a SX expression.  

Anyway, I'm also curious to see what the best implementation solution will be.

**Comment by Filippo-Tamagnini** [comment]
@alexandru-tc Feel free to move this in a separate discussion, if you think so.
To put it very simply: i have a differential equation that can be expressed as
_xdot = f(g(x))_
where _f(.)_ can be expressed easily in terms of primitive functions (thus as an SX functon).
_g(.)_, on the other hand is defined as a Callback (see: https://web.casadi.org/docs/#subclassing-callback ) and cannot take SX arguments, but only MX and DM. 
The thing only works if both x and xdot are expressed as MX, regardless of the fact that _g(f(x))_ is used in an ODE or PDE problem.

There is another catch: _g(x)_ requires (among other things) the computation of the eigenvalues of a tridiagonal matrix. If _g(x)_ appears in a discretized PDE problem, it's possible to call the eigenvalue routine only once on a block diagonal matrix. This is equivalent, in a way, to having a function _G(X)_ that is dependent on the values of _x_ at all grid points. 

Since the problem is quite involved and requires a quite low level use of CasADi, perhaps the simplest approach would be to leave the CasADi API exposed for the more advanced user, together with some utility functions to be used by the more novice.

**Comment by alexandru-tc** [comment]
I don't doubt the Callback prototype, that's what it is. What I'm curious is to know why you decided to implement like this, and what are the advantages. And this is why the discussion is better suited to a separate channel :)

**Comment by 4flixt** [comment]
Hi Filipo!

You got to the core of the problem there. 

> Perhaps it's possible to keep the variable list as a dictionary until the Model.setup method is called and only then create the structure.

This was also my first idea on how to approach the issue. The problem is, exactly as you said, that we want to have access to the variables before calling `model.setup()`. Having analogue SX variables might work for your problem but it is not what I would need when working with a MX model. In my use case I benefit from MX because it treats matrix-vector operations differently. As soon as I create such an expression with an SX variable the computation graph will be based on scalar operations. Compare for example the expressions below:

![Screenshot 2020-06-25 at 12 04 48](https://user-images.githubusercontent.com/32864011/85699875-2818dd80-b6dc-11ea-833e-a845ca042a4c.png)

This is more than just a preference. Especially when working with Neural Network System models (or other systems that require large matrix-vector operations), SX will require more memory and computation time as shown in this figure:

![Screenshot 2020-06-25 at 12 11 39](https://user-images.githubusercontent.com/32864011/85700931-27cd1200-b6dd-11ea-8570-945db814e6a7.png)

Note that in this investigation SX with more than 60 Neurons per Layer ran out of memory (hence the red area).

So I fear that for proper incorporation of MX (in my use case) all variables of the same kind (states, inputs, etc.) must be created at once. I haven't decided yet on the best API (ideally one that is similar to the current one for SX).

**Comment by Filippo-Tamagnini** [comment]
Right, that is why i referred to them as analogues, i'll make an example:


```
M = Model.model("continuous","MX")
n1 = ..
n2 = ..
A = DM( some matrix here )

X, x_a = M.set_variable("_x","MX_state",(n1,n2)) 
#This returns the MX variable (X of shape (n1,n2)) and its SX analogue (x_a of shape (n1,1))

expression_SX = ... here goes some expression of x_a
F = Function("F",[x_a],[expression_SX])
dX_dt = F(X) + A@X
M.set_rhs("MX_state",dX_dt)
```

**Comment by 4flixt** [comment]
I have to admit that I don't fully understand the idea of the analogous SX variable. The example also confuses me a bit. I guess `dX_dt = F(X) + A@X` is an arbitrary example but what is the point of 

```
expression_SX = ... here goes some expression of x_a
F = Function("F",[x_a],[expression_SX])
```

especially considering that (in your example) 

```
X, x_a = M.set_variable("_x","MX_state",(n1,n2)) 
```
already returns the MX variable `X`?

**Comment by 4flixt** [comment]
**I am now thinking of the following solution**:

We change the API of `model.set_variable` such that the input is `var_type` (e.g. `_x`, `_u`) and a list.
Each element of the list is a tuple consisting of `(var_name, shape)`. Here an example:

```python
x1, x2, x3 = model.set_variable('_x',[
    ('x_1', (2,)),
    ('x_2', (3,1)),
    ('x_3', (1,1)),
])
```

In this way we would not have to add variables to the MXStruct and everything else would work as previously intended (I hope).



This way the method allows to create multiple variables (of the same type) at once. This can be nice for the SX formulation but for MX it is required (!). In fact, I would lock the call for a given variable type after having defined them.

**Comment by Filippo-Tamagnini** [comment]
I guess this would do the trick, and it might also make the user-generated code more structured and readable. 

About my example, the point of the analogous variable would be to facilitate the definition of operations on vectors (or matrices) whose elements are homogeneous in type along at least one direction. 

Simple example: the state variable is the concentration of a substance in several points in space. The reaction rate can be generally expressed as a function of the local concentration (which is a scalar), however the transport term is better expressed with matrix operations.

My case is more complex, and the state variable is a matrix that stores the moments of a distribution for several points in space. In the first direction (rows), the elements of the matrix are inhomogenous in type (zeroth moment, first moment, second moment), while in the second direction they are homogenous (first grid point, second grid point, etc). The reaction term depends on all the values in one grid point, thus it would be nice to be able to define it in terms of a generic column vector.

Maybe i can illustrate this:
![immagine](https://user-images.githubusercontent.com/67159703/85729115-d03c9f80-b6f8-11ea-8e48-23e8feb4397c.png)
